lib/sqewer/local_connection.rb

```diff
@@ -0,0 +1,134 @@
+class Sqewer::LocalConnection < Sqewer::Connection
+  FAIL_AFTER_DELIVERIES = 10
+
+  def with_db(**k)
+    SQLite3::Database.open('sqewer-local-queue.sqlite3', **k) do |db|
+      db.busy_timeout = 5
+      return yield db
+    end
+  end
+
+  def with_readonly_db(&blk)
+    with_db(readonly: true, &blk)
+  end
+
+  def initialize(queue_url)
+    require 'sqlite3'
+    @queue_url = queue_url
+    with_db do |db|
+      db.execute("CREATE TABLE IF NOT EXISTS sqewer_messages_v1 (
+        id INTEGER PRIMARY KEY AUTOINCREMENT ,
+        queue_url VARCHAR NOT NULL,
+        receipt_handle VARCHAR NOT NULL,
+        deliver_after_epoch INTEGER,
+        times_delivered_so_far INTEGER DEFAULT 0,
+        last_delivery_at_epoch INTEGER,
+        visible BOOLEAN DEFAULT 't',
+        message_body TEXT)"
+      )
+      db.execute("CREATE INDEX IF NOT EXISTS index_sqewer_messages_v1_on_receipt_handle ON sqewer_messages_v1 (receipt_handle)")
+      db.execute("CREATE INDEX IF NOT EXISTS index_sqewer_messages_v1_on_queue_url ON sqewer_messages_v1 (queue_url)")
+    end
+  rescue LoadError => e
+    raise e, "You need the sqlite3 gem in your Gemfile to use LocalConnection. Add it to your Gemfile (`gem 'sqlite3'')"
+  end
+
+  # @return [Array<Message>] an array of Message objects 
+  def receive_messages
+    messages = load_receipt_handles_and_bodies
+    messages.map {|message| Message.new(message[0], message[1]) }
+  end
+
+  # @yield [#send_message] the object you can send messages through (will be flushed at method return)
+  # @return [void]
+  def send_multiple_messages
+    buffer = SendBuffer.new
+    yield(buffer)
+    messages = buffer.messages
+    persist_messages(messages)
+  end
+
+  # Deletes multiple messages after they all have been succesfully decoded and processed.
+  #
+  # @yield [#delete_message] an object you can delete an individual message through
+  # @return [void]
+  def delete_multiple_messages
+    buffer = DeleteBuffer.new
+    yield(buffer)
+    delete_persisted_messages(buffer.messages)
+  end
+
+  def truncate!
+    with_db do |db|
+      db.execute("DELETE FROM sqewer_messages_v1 WHERE queue_url = ?", @queue_url)
+    end
+  end
+
+  private
+
+  def delete_persisted_messages(messages)
+    ids_to_delete = messages.map{|m| m.fetch(:receipt_handle) }
+    with_db do |db|
+      db.execute("BEGIN")
+      ids_to_delete.each do |id|
+        db.execute("DELETE FROM sqewer_messages_v1 WHERE receipt_handle = ?", id)
+      end
+      db.execute("COMMIT")
```

Comment:
What happens when an exception occurs? Do we ever release the transaction? Does the transaction complete partially?

This is relevant for several other methods as well.

Reply:
If an exception happens the transaction is not going to commit, and the exception will be raised up to the caller. But given that this is not meant for production use, I think we can cut a few corners (don't even start on the race conditions here :-P)